src/main/java/cubicchunks/world/provider/VanillaCubicProvider.java

```diff
@@ -0,0 +1,384 @@
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in
+ *  all copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ *  THE SOFTWARE.
+ */
+package cubicchunks.world.provider;
+
+import javax.annotation.Nullable;
+
+import cubicchunks.world.ICubicWorld;
+import cubicchunks.world.provider.CubicWorldProvider;
+import cubicchunks.world.type.ICubicWorldType;
+import cubicchunks.worldgen.generator.ICubeGenerator;
+import cubicchunks.worldgen.generator.vanilla.VanillaCompatibilityGenerator;
+import net.minecraft.entity.player.EntityPlayerMP;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Vec3d;
+import net.minecraft.world.DimensionType;
+import net.minecraft.world.World;
+import net.minecraft.world.WorldProvider;
+import net.minecraft.world.WorldType;
+import net.minecraft.world.biome.Biome;
+import net.minecraft.world.biome.BiomeProvider;
+import net.minecraft.world.border.WorldBorder;
+import net.minecraft.world.chunk.IChunkGenerator;
+import net.minecraftforge.fml.relauncher.Side;
+import net.minecraftforge.fml.relauncher.SideOnly;
+
+public class VanillaCubicProvider extends CubicWorldProvider {
+
+	private static final WorldType HEURISTIC_WORLDTYPE = new WorldType("no-op") {
+		@Override public IChunkGenerator getChunkGenerator(World world, String options) {
+			return null; // now thats something we can detect >:D
+		}
+	};
+
+	private WorldProvider provider;
+
+	private ICubeGenerator cubeGen;
+
+	public VanillaCubicProvider(ICubicWorld world, WorldProvider provider, @Nullable IChunkGenerator reUse) {
+		this.provider = provider;
+		this.worldObj = (World) world;
+
+		boolean useProvider = false;
+
+		if(worldObj.getWorldType() instanceof ICubicWorldType) { // Who do we trust!??!?! D:
+
+			// nasty hack heuristic to see if provider asks its WorldType for a chunk generator
+			//ReflectionUtil.setFieldValueSrg(wp, "field_76577_b", HEURISTIC_WORLDTYPE);
+			provider.terrainType = HEURISTIC_WORLDTYPE;
+
+			IChunkGenerator pro_or_null = provider.createChunkGenerator();
+
+			// clean up
+			//ReflectionUtil.setFieldValueSrg(provider, "field_76577_b", worldObj.getWorldType());
+			provider.terrainType = HEURISTIC_WORLDTYPE;
```

Comment:
When I removed the ReflectionUtil method, I made a mistake...
(The second terrainType = should have been worldObj.getWorldType()... fixing)